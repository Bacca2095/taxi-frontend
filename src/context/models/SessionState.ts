export interface SessionState {
  document?: string;
  sessionId?: string;
}
