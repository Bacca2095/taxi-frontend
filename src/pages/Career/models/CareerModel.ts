/* eslint-disable prettier/prettier */
export interface CareerModel {
  id?: number;
  nombre: string;
  documento: string;
  telefono: number;
  fechaRecogida: string | Date;
  horaRecogida: string;
  direccion: string;
  costo?: number;
}
