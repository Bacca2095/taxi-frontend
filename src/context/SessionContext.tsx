/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React, { createContext, useEffect, useRef, useState } from 'react';
import { SessionState } from 'context/models/SessionState';
import * as sessionService from '../services/sessionService';

export const useStateContainer = (
  initialState: SessionState = { document: '', sessionId: '' },
) => {
  const [document, setDocument] = useState(initialState.document);
  const [sessionId, setSessionId] = useState(initialState.sessionId);
  const documentRef = useRef(false);

  const clearSession = () => {
    sessionService.deleteSession();
  };

  useEffect(() => {
    if (!documentRef.current) {
      documentRef.current = true;
      return;
    }
    if (!document) {
      return;
    }
    setSessionId(sessionService.createSession(document));
  }, [document]);

  return {
    data: { document, sessionId },
    mutations: { setDocument, clearSession },
  };
};

export const SessionContext = createContext<
  ReturnType<typeof useStateContainer>
>({} as never);

export const SessionProvider: React.FC = ({ children }) => {
  const contextValue = useStateContainer(sessionService.recoverSession());
  return (
    <SessionContext.Provider value={contextValue}>
      {children}
    </SessionContext.Provider>
  );
};
