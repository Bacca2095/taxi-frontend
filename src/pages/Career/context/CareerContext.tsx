/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React, {
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import { SessionContext } from 'context/SessionContext';
import * as careerService from '../services/careerService';
import { CareerModel } from '../models/CareerModel';

export interface CareerState {
  careers?: CareerModel[];
}

export const useStateContainer = (
  initialState: CareerState = {},
  sessionContext: typeof SessionContext,
) => {
  const {
    data: { sessionId },
  } = useContext(sessionContext);
  const [allCareers, setAllCareers] = useState(initialState.careers || []);
  const [currentCareerId, setCurrentCareerId] = useState(0);
  const [errorOnDelete, setErrorOnDelete] = useState(false);
  const [errorOnCreate, setErrorOnCreate] = useState(false);
  const [updateList, setUpdateList] = useState(false);
  const careerRef = useRef(false);

  const deleteCareer = () => {
    careerService.deleteCareer(currentCareerId).then((res) => {
      if (res) {
        setErrorOnDelete(false);
      } else {
        setErrorOnDelete(true);
      }
      setUpdateList(!updateList);
    });
  };

  const createCareer = (career: CareerModel) => {
    careerService.createCareer(career).then((res) => {
      if (res) {
        setErrorOnCreate(false);
      } else {
        setErrorOnCreate(true);
      }

      setUpdateList(!updateList);
    });
  };

  useEffect(() => {
    if (!careerRef.current) {
      careerRef.current = true;
      return;
    }
    if (!document) {
      return;
    }
    careerService
      .listCareer(sessionId?.split(':::')[1])
      .then((career) => setAllCareers(career));
  }, [sessionId, updateList]);

  return {
    data: { allCareers, currentCareerId, errorOnDelete, errorOnCreate },
    mutations: {
      setCurrentCareerId,
      deleteCareer,
      createCareer,
    },
  };
};

export const CareerContext = createContext<
  ReturnType<typeof useStateContainer>
>({} as never);

export interface CareerProviderProps {
  initialState?: CareerState;
}
export const CareerProvider: React.FC<CareerProviderProps> = ({
  children,
  initialState = {},
}) => {
  const contextValue = useStateContainer(initialState, SessionContext);
  return (
    <CareerContext.Provider value={contextValue}>
      {children}
    </CareerContext.Provider>
  );
};
