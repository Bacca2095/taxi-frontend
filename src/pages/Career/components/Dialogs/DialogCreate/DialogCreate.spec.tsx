import React from 'react';
import nock from 'nock';
import { render, fireEvent, waitFor, act } from '@testing-library/react';
import { SessionContext } from 'context/SessionContext';
import { CareerProvider } from 'pages/Career/context/CareerContext';
import * as careerFixture from 'tests/fixtures/career.fixture';
import { CareerModel } from 'pages/Career/models/CareerModel';
import { DialogCreate } from '.';

describe('Dialog Create tests', () => {
  let onClose: jest.Mock;
  let onCreate: (career: CareerModel) => void;
  let setDocument: jest.Mock;
  let clearSession: jest.Mock;
  let careers: CareerModel[];
  let career: CareerModel;
  const document = '12345';
  const sessionId = '12345:::12345';

  beforeAll(() => {
    onClose = jest.fn();
    setDocument = jest.fn();
    clearSession = jest.fn();
    career = careerFixture.getSingle();
    careers = careerFixture.getList();
    onCreate = jest.fn();
    nock('http://localhost:3001/api')
      .persist()
      .get('/carreras/12345')
      .reply(200, careers)
      .post('/carreras')
      .reply(201, {});
  });

  it('should match snapshot', async () => {
    const { container } = render(
      <SessionContext.Provider
        value={{
          data: { document, sessionId },
          mutations: { setDocument, clearSession },
        }}
      >
        <CareerProvider>
          <DialogCreate open onClose={onClose} onCreate={onCreate} />
        </CareerProvider>
      </SessionContext.Provider>,
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('should test not save', async () => {
    const { findByTestId, findAllByText } = render(
      <SessionContext.Provider
        value={{
          data: { document, sessionId },
          mutations: { setDocument, clearSession },
        }}
      >
        <CareerProvider>
          <DialogCreate open onClose={onClose} onCreate={onCreate} />
        </CareerProvider>
      </SessionContext.Provider>,
    );

    const name = await findByTestId('input-name');
    const phone = await findByTestId('input-phone');
    const address = await findByTestId('input-address');
    const save = await findByTestId('dialog-create-submit');

    fireEvent.change(name, { target: { value: career.nombre } });
    fireEvent.change(phone, { target: { value: career.telefono } });
    fireEvent.change(address, { target: { value: career.direccion } });

    fireEvent.change(name, { target: { value: '' } });
    fireEvent.change(phone, { target: { value: '' } });
    fireEvent.change(address, { target: { value: '' } });

    fireEvent.click(save);

    const count = await findAllByText(/obligatorio/i);

    expect(count.length).toBe(3);
  });

  it('should test save', async () => {
    const { findByTestId } = render(
      <SessionContext.Provider
        value={{
          data: { document, sessionId },
          mutations: { setDocument, clearSession },
        }}
      >
        <CareerProvider>
          <DialogCreate open onClose={onClose} onCreate={onCreate} />
        </CareerProvider>
      </SessionContext.Provider>,
    );

    const name = await findByTestId('input-name');
    const phone = await findByTestId('input-phone');
    const address = await findByTestId('input-address');
    const date = await findByTestId('input-date');

    fireEvent.change(name, { target: { value: career.nombre } });
    fireEvent.change(phone, { target: { value: career.telefono } });
    fireEvent.change(address, { target: { value: career.direccion } });
    fireEvent.change(date, { target: { value: career.fechaRecogida } });

    const save = await findByTestId('dialog-create-submit');

    await act(async () => {
      fireEvent.click(save);
    });

    await waitFor(() => {
      expect(onCreate).toHaveBeenCalled();
    });
  });
});
