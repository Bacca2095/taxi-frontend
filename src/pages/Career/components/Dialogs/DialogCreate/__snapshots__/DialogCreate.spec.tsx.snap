// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Create tests should match snapshot 1`] = `
<div
  aria-hidden="true"
>
  <form>
    <div
      class="MuiDialog-root"
      role="presentation"
      style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper"
        role="none presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
          role="dialog"
        >
          <div
            class="MuiDialogTitle-root"
            color="primary"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4"
            >
              Solicitar Carrera
            </h4>
          </div>
          <div
            class="MuiDialogContent-root MuiDialogContent-dividers"
          >
            <p
              class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
            >
              Complete los datos para solicitar la carrera
            </p>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-center"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                    data-shrink="false"
                  >
                    Nombre
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input"
                      data-testid="input-name"
                      name="nombre"
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-3"
                      >
                        <span>
                          Nombre
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                    data-shrink="false"
                  >
                    Teléfono
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input"
                      data-testid="input-phone"
                      name="telefono"
                      type="number"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-3"
                      >
                        <span>
                          Teléfono
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                    data-shrink="false"
                  >
                    Dirección
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input"
                      data-testid="input-address"
                      name="direccion"
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-3"
                      >
                        <span>
                          Dirección
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                    data-shrink="true"
                  >
                    Fecha
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                      data-testid="input-date"
                      type="text"
                      value="2021/09/06"
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                    >
                      <button
                        class="MuiButtonBase-root MuiIconButton-root"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                            />
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
                      >
                        <span>
                          Fecha
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                  
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                    data-shrink="true"
                  >
                    Hora
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                      data-testid="input-time"
                      type="text"
                      value="11:40 PM"
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                    >
                      <button
                        class="MuiButtonBase-root MuiIconButton-root"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                            />
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                            />
                          </svg>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
                      >
                        <span>
                          Hora
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                  
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              data-testid="dialog-create-cancel"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Cancelar
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
              data-testid="dialog-create-submit"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Solicitar
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </form>
</div>
`;
